### PR TITLE
Add format options for ansible-galaxy collection list

### PR DESCRIPTION
**Pull request by shanemcd**
##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME
`ansible-galaxy`

##### ADDITIONAL INFORMATION
This would be useful for an upcoming feature in AWX/Tower.

### CODE REVIEW

File: lib/ansible/cli/galaxy.py [hunk 1]
```diff
@@ -301,6 +302,10 @@ def add_list_options(self, parser, parents=None):
 
         list_parser.add_argument(galaxy_type, help=galaxy_type.capitalize(), nargs='?', metavar=galaxy_type)
 
+        if galaxy_type == 'collection':
+            list_parser.add_argument('-o', '--output', dest='output_format',
```
Comment (jborean93) [inline lib/ansible/cli/galaxy.py]: might be a good idea to have a 3rd option like `default` or `human` which is the current behaviour. You can also have a `choices=[]` kwarg where you define the valid choices. This helps both on validation and people using argcomplete to tab complete the options.

**Review by jborean93 — CHANGES_REQUESTED** [review]
I'm fine with the idea, we just need a changelog fragment and integration test for this. You can find the current tests at https://github.com/ansible/ansible/blob/devel/test/integration/targets/ansible-galaxy-collection/tasks/list.yml.

File: lib/ansible/cli/galaxy.py [hunk 2]
```diff
@@ -1313,7 +1318,8 @@ def execute_list(self):
         if context.CLIARGS['type'] == 'role':
             self.execute_list_role()
         elif context.CLIARGS['type'] == 'collection':
-            self.execute_list_collection()
+            output_format = context.CLIARGS.get('output_format', False)
```
Comment (webknjaz) [inline lib/ansible/cli/galaxy.py]: This should be moved inside `execute_list_collection()` for consistency — all other code accesses `context.CLIARGS` directly so it's unnecessary to have an injection here.

File: lib/ansible/cli/galaxy.py [hunk 3]
```diff
@@ -1386,6 +1392,9 @@ def execute_list_collection(self, artifacts_manager=None):
         warnings = []
         path_found = False
         collection_found = False
+        if output_format is not None:
+            collections_in_paths = {}
```
Comment (webknjaz) [inline lib/ansible/cli/galaxy.py]: Declaring a variable in only one of the code branches is dangerous because it's hard to maintain accessing it under the same conditions in other places which can become a real bug during refactoring.
```suggestion
        collections_in_paths = {}
```

File: lib/ansible/cli/galaxy.py [hunk 4]
```diff
@@ -1451,13 +1460,18 @@ def execute_list_collection(self, artifacts_manager=None):
                     display.vvv("No collections found at {0}".format(collection_path))
                     continue
 
-                # Display header
-                fqcn_width, version_width = _get_collection_widths(collections)
-                _display_header(collection_path, 'Collection', 'Version', fqcn_width, version_width)
+                if output_format is not None:
```
Comment (webknjaz) [inline lib/ansible/cli/galaxy.py]: I think that the first branch of `if collection_name` should also emit JSON. Also, with Jordan's suggestion to use the `choices` kwarg in the CLI declaration, it'd be something like `output_format == 'json'`.

File: lib/ansible/cli/galaxy.py [hunk 5]
```diff
@@ -1466,6 +1480,14 @@ def execute_list_collection(self, artifacts_manager=None):
         for w in warnings:
             display.warning(w)
 
+        if output_format is not None:
+            if output_format == 'json':
+                display.display(json.dumps(collections_in_paths))
+            elif output_format == 'yaml':
+                display.display(yaml.safe_dump(collections_in_paths))
+            else:
+                raise AnsibleError("Only yaml or json are supported output types")
```
Comment (webknjaz) [inline lib/ansible/cli/galaxy.py]: This could be less nested once you apply `choices`.

File: lib/ansible/cli/galaxy.py [hunk 6]
```diff
@@ -12,6 +12,7 @@
 import textwrap
 import time
 import yaml
+import json
```
Comment (webknjaz) [inline lib/ansible/cli/galaxy.py]: This is injected right between two `yaml` imports which is also a third-party. Please move it up, before line 8 (all other imports are sorted alphabetically).

**Review by webknjaz — CHANGES_REQUESTED** [review]
This needs integration tests + some structural improvements:

File: lib/ansible/cli/galaxy.py (hunk 4 above)
Comment (shanemcd) [reply lib/ansible/cli/galaxy.py]: Not sure I'm following. We loop over `collections_search_paths`. We need to include all of these in the json that's emitted, so we can only do that once we have the entirety of the data that needs to be printed.

**Comment by shanemcd** [comment]
@jborean93 @webknjaz I have addressed the majority of your feedback. However, I have been unsuccessful at getting the integration tests to work on my machine, so I cannot add a new one.

Using the command:

```
bin/ansible-test integration ansible-galaxy-collection -v
```

I see the error:

```
TASK [ansible-galaxy-collection : install simple collection with implicit path - galaxy_ng] **************************************************************************************************************
fatal: [testhost]: FAILED! => {"changed": true, "cmd": ["ansible-galaxy", "collection", "install", "namespace1.name1", "-s", "galaxy_ng", "-v"], "delta": "0:00:01.617742", "end": "2021-02-04 10:04:31.690424", "msg": "non-zero return code", "rc": 1, "start": "2021-02-04 10:04:30.072682", "stderr": "\u001b[0;31mERROR! Failed to download collection tar from 'galaxy_ng': <urlopen error [Errno -2] Name or service not known>\u001b[0m", "stderr_lines": ["\u001b[0;31mERROR! Failed to download collection tar from 'galaxy_ng': <urlopen error [Errno -2] Name or service not known>\u001b[0m"], "stdout": "\u001b[0;34mUsing /tmp/ansible.xmzexh7v.test/galaxy/ansible.cfg as config file\u001b[0m\nStarting galaxy collection install process\nProcess install dependency map\nStarting collection install process\nDownloading http://ansible-ci-pulp:80/pulp/content/published/namespace1-name1-1.0.9.tar.gz to /home/shanemcd/.ansible/tmp/ansible-local-139077gi10ktw5/tmpr1mb62q0/namespace1-name1-1.0.9-aer09auq", "stdout_lines": ["\u001b[0;34mUsing /tmp/ansible.xmzexh7v.test/galaxy/ansible.cfg as config file\u001b[0m", "Starting galaxy collection install process", "Process install dependency map", "Starting collection install process", "Downloading http://ansible-ci-pulp:80/pulp/content/published/namespace1-name1-1.0.9.tar.gz to /home/shanemcd/.ansible/tmp/ansible-local-139077gi10ktw5/tmpr1mb62q0/namespace1-name1-1.0.9-aer09auq"]}
```

I spent about an hour trying to figure this out and had 0 luck.

**Comment by s-hertel** [comment]
These integration tests need to be run with --docker

File: test/integration/targets/ansible-galaxy-collection/tasks/list.yml [hunk 7]
```diff
@@ -31,6 +31,28 @@
...
 
+- name: list collections in yaml format
+  command: ansible-galaxy collection list -o yaml
+  register: list_result_yaml
+  environment:
+    ANSIBLE_COLLECTIONS_PATH: "{{ galaxy_dir }}/dev:{{ galaxy_dir }}/prod"
+
+- assert:
+    that:
+      - "{{ item.value | length == 3 }}"
```
Comment (webknjaz) [inline test/integration/targets/ansible-galaxy-collection/tasks/list.yml]: You don't need to render this through "mustache"
```suggestion
    - item.value | length == 3
```

File: test/integration/targets/ansible-galaxy-collection/tasks/list.yml [hunk 8]
```diff
@@ -31,6 +31,28 @@
...
+
+- name: list collections in json format
+  command: ansible-galaxy collection list -o json
+  register: list_result_json
+  environment:
+    ANSIBLE_COLLECTIONS_PATH: "{{ galaxy_dir }}/dev:{{ galaxy_dir }}/prod"
+
+- assert:
+    that:
+      - "{{ item.value | length == 3 }}"
```
Comment (webknjaz) [inline test/integration/targets/ansible-galaxy-collection/tasks/list.yml:59]: I'd expect the test to check the contents, not just length.

File: test/integration/targets/ansible-galaxy-collection/tasks/list.yml [hunk 9]
```diff
@@ -31,6 +31,28 @@
       - "'dev.collection2 placeholder' in list_result.stdout"
       - "'dev.collection3 *' in list_result.stdout"
 
+- name: list collections in yaml format
+  command: ansible-galaxy collection list -o yaml
```
Comment (webknjaz) [inline test/integration/targets/ansible-galaxy-collection/tasks/list.yml]: I'd expect extra tests for listing specific collections too. Like `ansible-galaxy collection list -o yaml ns1.coll2` and `ansible-galaxy collection list -o yaml ns1.coll2:1.0` maybe...

File: lib/ansible/cli/galaxy.py [hunk 10]
```diff
@@ -1379,9 +1384,11 @@ def execute_list_collection(self, artifacts_manager=None):
         :param artifacts_manager: Artifacts manager.
         """
 
+        output_format = context.CLIARGS.get('output_format', False)
```
Comment (webknjaz) [inline lib/ansible/cli/galaxy.py]: Since `default` + `choices` are used, it's guaranteed that the key exists and so there's no point in defaulting to `False`. There's no way it's useful. So let's not assume a missing key. If somebody refactors things and it'll start failing, then this must start failing loudly too.
```suggestion
        output_format = context.CLIARGS['output_format']
```

File: lib/ansible/cli/galaxy.py [hunk 11]
```diff
@@ -1451,13 +1458,18 @@ def execute_list_collection(self, artifacts_manager=None):
                     display.vvv("No collections found at {0}".format(collection_path))
                     continue
 
-                # Display header
-                fqcn_width, version_width = _get_collection_widths(collections)
-                _display_header(collection_path, 'Collection', 'Version', fqcn_width, version_width)
+                if output_format in ('yaml', 'json'):
```
Comment (webknjaz) [inline lib/ansible/cli/galaxy.py]: Semantically, it's better to use a set here:
```suggestion
                if output_format in {'yaml', 'json'}:
```

File: lib/ansible/cli/galaxy.py [hunk 12]
```diff
@@ -1451,13 +1458,18 @@ def execute_list_collection(self, artifacts_manager=None):
...
+                    collections_in_paths[collection_path] = [
```
Comment (webknjaz) [inline lib/ansible/cli/galaxy.py]: Why add another level of nesting with a list? Just have a dict comprehension.

File: lib/ansible/cli/galaxy.py (hunk 4 above)
Comment (webknjaz) [reply lib/ansible/cli/galaxy.py]: Just think about `ansible-galaxy collection list ns.name -o yaml`. Currently, your logic skips this case and falls back to the human-readable format.

File: lib/ansible/cli/galaxy.py [hunk 13]
```diff
@@ -1451,13 +1458,18 @@ def execute_list_collection(self, artifacts_manager=None):
...
+                    # Display header
+                    fqcn_width, version_width = _get_collection_widths(collections)
+                    _display_header(collection_path, 'Collection', 'Version', fqcn_width, version_width)
 
-                # Sort collections by the namespace and name
-                for collection in sorted(collections, key=to_text):
-                    _display_collection(collection, fqcn_width, version_width)
+                    # Sort collections by the namespace and name
+                    for collection in sorted(collections, key=to_text):
+                        _display_collection(collection, fqcn_width, version_width)
```
Comment (webknjaz) [inline lib/ansible/cli/galaxy.py]: Please dedent this back to keep the nesting sane. It's better to use `continue` to skip it when necessary instead.
```suggestion
                    continue

                # Display header
                fqcn_width, version_width = _get_collection_widths(collections)
                _display_header(collection_path, 'Collection', 'Version', fqcn_width, version_width)

                # Sort collections by the namespace and name
                for collection in sorted(collections, key=to_text):
                    _display_collection(collection, fqcn_width, version_width)
```

**Review by webknjaz — CHANGES_REQUESTED** [review]
Looking at the source, I'm pretty sure that `ansible-galaxy collection list ns.name -o json` will output a human-readable table followed by an empty JSON object `{}`. Please add a test for this and check that only JSON is returned.

File: lib/ansible/cli/galaxy.py (hunk 12 above)
Comment (jborean93) [reply lib/ansible/cli/galaxy.py]: In case we ever add more field to this like author name or something of the sort. Keeping it as a dict entry in a list gives us more flexibility in the future.

File: lib/ansible/cli/galaxy.py (hunk 10 above)
Comment (jborean93) [reply lib/ansible/cli/galaxy.py]: Agree on this it should change.

**Review by jborean93 — APPROVED** [review]
Apart from some of the comments already made by @webknjaz I'm fairly happy with this PR.

File: test/integration/targets/ansible-galaxy-collection/tasks/list.yml (hunk 9 above)
Comment (shanemcd) [reply test/integration/targets/ansible-galaxy-collection/tasks/list.yml]: Added below

File: lib/ansible/cli/galaxy.py (hunk 10 above)
Comment (shanemcd) [reply lib/ansible/cli/galaxy.py]: Done.

File: lib/ansible/cli/galaxy.py (hunk 12 above)
Comment (shanemcd) [reply lib/ansible/cli/galaxy.py]: After further consideration, I actually agree with this point. It makes looping over this data easier too, as seen in the tests. If we wanted to add more information to the sub-dictionary in the future, that's possible too. Thanks for this feedback.

File: lib/ansible/cli/galaxy.py (hunk 13 above)
Comment (shanemcd) [reply lib/ansible/cli/galaxy.py]: Done.

**Comment by shanemcd** [comment]
> Looking at the source, I'm pretty sure that `ansible-galaxy collection list ns.name -o json` will output a human-readable table followed by an empty JSON object `{}`. Please add a test for this and check that only JSON is returned.

Not sure how I overlooked that branch in the code. I've fixed this and added some tests. Thanks.

File: lib/ansible/cli/galaxy.py (hunk 11 above)
Comment (shanemcd) [reply lib/ansible/cli/galaxy.py]: Done.

File: lib/ansible/cli/galaxy.py (hunk 12 above)
Comment (webknjaz) [reply lib/ansible/cli/galaxy.py]: Well, if we want flexibility, we could have a dict value but I still expect it to we inside a dict, not inside a list.
It's `{'ns.coll': {'fqcn': 'ns.coll', 'ver': '1.2.0'}}` vs `[{'fqcn': 'ns.coll', 'ver': '1.2.0'}]`: with a list, it's harder to access a specific collection as you'd have to do a full list scan to find in which element your target is while with a dict you can access it by key instantly.

cc @shanemcd

Comment (webknjaz) [reply lib/ansible/cli/galaxy.py]: Oh, I see it's addressed now.

File: lib/ansible/cli/galaxy.py [hunk 14]
```diff
@@ -1430,6 +1437,13 @@ def execute_list_collection(self, artifacts_manager=None):
 
                 fqcn_width, version_width = _get_collection_widths([collection])
 
+                if output_format in {'yaml', 'json'}:
+                    collections_in_paths[collection_path] = {
+                        collection.fqcn: { 'version': collection.ver }
+                    }
+
+                    continue
```
Comment (webknjaz) [inline lib/ansible/cli/galaxy.py:1443]: Please move this above the `_get_collection_widths()` call to avoid running that code unnecessarily. It'd be basically the same as what's done in the else-branch that has a similar code below this snippet.

File: test/integration/targets/ansible-galaxy-collection/tasks/list.yml [hunk 15]
```diff
@@ -31,6 +31,58 @@
...
+      - "item.value['dev.collection2'].version == 'placeholder'"
+      - "item.value['dev.collection3'].version == '*'"
+  with_dict: "{{ list_result_json.stdout | from_json }}"
+
+- name: list single collection in json format
+  command: "ansible-galaxy collection list {{ item.key }} -o json"
+  register: list_single_result_json
+  environment:
+    ANSIBLE_COLLECTIONS_PATH: "{{ galaxy_dir }}/dev:{{ galaxy_dir }}/prod"
+  with_dict: "{{ { 'dev.collection1': '*', 'dev.collection2': 'placeholder', 'dev.collection3': '*' } }}"
```
Comment (webknjaz) [inline test/integration/targets/ansible-galaxy-collection/tasks/list.yml]: This is rather overengineered:
```suggestion
  command: "ansible-galaxy collection list {{ item }} -o json"
  register: list_single_result_json
  environment:
    ANSIBLE_COLLECTIONS_PATH: "{{ galaxy_dir }}/dev:{{ galaxy_dir }}/prod"
  loop:
  - dev.collection1
  - dev.collection2
  - dev.collection3
```

**Review by webknjaz — CHANGES_REQUESTED** [review]
Almost LGTM: I have a few final comments below.

Also, could you please add tests with explicit `ansible-galaxy collection list -o human`? I believe I've only seen an implicit one.

**Review by s-hertel — APPROVED** [review]
Nice addition!

File: lib/ansible/cli/galaxy.py [hunk 16]
```diff
@@ -301,6 +302,10 @@ def add_list_options(self, parser, parents=None):
 
         list_parser.add_argument(galaxy_type, help=galaxy_type.capitalize(), nargs='?', metavar=galaxy_type)
 
+        if galaxy_type == 'collection':
+            list_parser.add_argument('-o', '--output', dest='output_format', choices=('human', 'yaml', 'json'), default='human',
```
Comment (bcoca) [inline lib/ansible/cli/galaxy.py]: we use -o for output file elsewhere, it is not good to overload options with diff meanings

Comment (webknjaz) [reply lib/ansible/cli/galaxy.py]: Fair enough. How about `-f` / `--output-format`?

Comment (s-hertel) [reply lib/ansible/cli/galaxy.py]: Could use `--format` to be similar to pip. I wouldn't add `-f` though since it's already associated with `--force`.

File: test/integration/targets/ansible-galaxy-collection/tasks/list.yml (hunk 15 above)
Comment (shanemcd) [reply test/integration/targets/ansible-galaxy-collection/tasks/list.yml]: Sorry if it seems that way. But if you look at the assertions below, I am using the value in this dictionary to access the expected version of the collection. If you have a suggestion on how to better restructure this test, I am happy to update. But this suggestion breaks the tests.

File: lib/ansible/cli/galaxy.py (hunk 16 above)
Comment (webknjaz) [reply lib/ansible/cli/galaxy.py]: Right, missed that. How about `--output-as`?

File: lib/ansible/cli/galaxy.py (hunk 14 above)
Comment (shanemcd) [reply lib/ansible/cli/galaxy.py:1443]: Done

File: lib/ansible/cli/galaxy.py (hunk 16 above)
Comment (shanemcd) [reply lib/ansible/cli/galaxy.py]: I think `--format` makes sense. Let's go with that. Thanks all.

File: test/integration/targets/ansible-galaxy-collection/tasks/list.yml (hunk 15 above)
Comment (webknjaz) [reply test/integration/targets/ansible-galaxy-collection/tasks/list.yml]: @shanemcd I'm pretty it shouldn't break tests in this place. You don't use the values. The assertion below uses the results that come from the stdout output. In this specific task, you only use keys.

**Comment by shanemcd** [comment]
> Almost LGTM: I have a few final comments below.
> 
> Also, could you please add tests with explicit `ansible-galaxy collection list -o human`? I believe I've only seen an implicit one.

This is done.

**Review by webknjaz — APPROVED** [review]

File: test/integration/targets/ansible-galaxy-collection/tasks/list.yml (hunk 15 above)
Comment (shanemcd) [reply test/integration/targets/ansible-galaxy-collection/tasks/list.yml]: @webknjaz I am using the values in the assertion directly proceeding this task. Notice ` == item.item.value`

Comment (webknjaz) [reply test/integration/targets/ansible-galaxy-collection/tasks/list.yml]: Yes, but that refers to what you feed to the assertion task's `with_dict`, nothing to do with this task.

Comment (webknjaz) [reply test/integration/targets/ansible-galaxy-collection/tasks/list.yml]: Oh... I see what you did. You use a closure. This is clever but it is also hard to follow.

**Review by bcoca — APPROVED** [review]

File: lib/ansible/cli/galaxy.py [hunk 17]
```diff
@@ -301,6 +302,10 @@ def add_list_options(self, parser, parents=None):
 
         list_parser.add_argument(galaxy_type, help=galaxy_type.capitalize(), nargs='?', metavar=galaxy_type)
 
+        if galaxy_type == 'collection':
+            list_parser.add_argument('--format', dest='output_format', choices=('human', 'yaml', 'json'), default='human',
```
Comment (samdoran) [inline lib/ansible/cli/galaxy.py:306]: Do we need to specify `human` as an option? That could just the default and `yaml` or `json` are specified.

File: lib/ansible/cli/galaxy.py [hunk 18]
```diff
@@ -1466,6 +1487,11 @@ def execute_list_collection(self, artifacts_manager=None):
         for w in warnings:
             display.warning(w)
 
+        if output_format == 'json':
+            display.display(json.dumps(collections_in_paths))
+        elif output_format == 'yaml':
+            display.display(yaml.safe_dump(collections_in_paths))
+
```
Comment (samdoran) [inline lib/ansible/cli/galaxy.py:1497]: If no collection paths are found, this will output an empty `{}` before the help and error message. This should be after the `raise AnsibleOptionsError` below.

```
> ANSIBLE_COLLECTIONS_PATH='' ansible-galaxy collection list --format json
{}
usage: ansible-galaxy [-h] [--version] [-v] TYPE ...

Perform various Role and Collection related operations.

positional arguments:
  TYPE
    collection   Manage an Ansible Galaxy collection.
    role         Manage an Ansible Galaxy role.

optional arguments:
  --version      show program's version number, config file location, configured module search path, module location, executable location and exit
  -h, --help     show this help message and exit
  -v, --verbose  verbose mode (-vvv for more, -vvvv to enable connection debugging)
ERROR! - None of the provided paths were usable. Please specify a valid path with --collections-path
```

**Review by samdoran — CHANGES_REQUESTED** [review]

File: lib/ansible/cli/galaxy.py (hunk 17 above)
Comment (samdoran) [reply lib/ansible/cli/galaxy.py:306]: This is not a blocker, just a thought on the interface.

Comment (shanemcd) [reply lib/ansible/cli/galaxy.py:306]: @samdoran This was requested here: https://github.com/ansible/ansible/pull/73474#discussion_r569948199 I have also added a test for this functionality.

 I don't have a strong opinion either way. I will defer to the group. @s-hertel @webknjaz @jborean93 @bcoca - thoughts?

File: lib/ansible/cli/galaxy.py (hunk 18 above)
Comment (webknjaz) [reply lib/ansible/cli/galaxy.py:1497]: Why? Doesn't it go to stdout while help goes to stderr? With typical piping, it'd be alright and the tools down the chain would be able to consume and empty "there's nothing installed" `{}`...

File: lib/ansible/cli/galaxy.py (hunk 17 above)
Comment (webknjaz) [reply lib/ansible/cli/galaxy.py:306]: I was unable to find examples of a `default` that is not listed in `choices` on the internet. This may need to be checked experimentally.

Comment (webknjaz) [reply lib/ansible/cli/galaxy.py:306]: @shanemcd my understanding is that @sdoran suggests
```suggestion
            list_parser.add_argument('--format', dest='output_format', choices=('yaml', 'json'), default='human',
```
and it's undocumented whether this would work argparse-wise.

File: lib/ansible/cli/galaxy.py (hunk 18 above)
Comment (shanemcd) [reply lib/ansible/cli/galaxy.py:1497]: Awesome find. Fixed + added test.

Comment (shanemcd) [reply lib/ansible/cli/galaxy.py:1497]: The empty JSON object was being included in stdout along with the usage text. IMO this _was_ a bug / undesirable behavior.

Comment (samdoran) [reply lib/ansible/cli/galaxy.py:1497]: I didn't know if it was intentional to output an empty dict. I don't think it will cause issues, just seems like it shouldn't be doing that.

Comment (shanemcd) [reply lib/ansible/cli/galaxy.py:1497]: Demonstration:

```
$ ANSIBLE_COLLECTIONS_PATH='' ansible-galaxy collection list --format yaml 2> /dev/null
{}
usage: ansible-galaxy [-h] [--version] [-v] TYPE ...

Perform various Role and Collection related operations.

positional arguments:
  TYPE
    collection   Manage an Ansible Galaxy collection.
    role         Manage an Ansible Galaxy role.

optional arguments:
  --version      show program's version number, config file location, configured module search path, module location, executable location and exit
  -h, --help     show this help message and exit
  -v, --verbose  verbose mode (-vvv for more, -vvvv to enable connection debugging)
```

File: lib/ansible/cli/galaxy.py (hunk 17 above)
Comment (shanemcd) [reply lib/ansible/cli/galaxy.py:306]: It _does_ seem to work. Again, I have no strong feelings here. I'm happy to make this change + delete the explicit 'human' format test if that's what folks want.

File: lib/ansible/cli/galaxy.py (hunk 18 above)
Comment (shanemcd) [reply lib/ansible/cli/galaxy.py:1497]: Some options I can identify:

- Leave it as-is (patch w/ Sam's suggestion)
- We stop using `AnsibleOptionsError` here: https://github.com/ansible/ansible/blob/649df97d9174a285bcb08b881c45d2af4058d450/lib/ansible/cli/galaxy.py#L1491
- Make `AnsibleOptionsError` send to stderr
- We change the conditional that throws the error to be `if not path_found and output_format != 'human'`.

Comment (samdoran) [reply lib/ansible/cli/galaxy.py:1497]: The use of `AnsibleOptionsError` is to make in consistent with other sub commands. 

It looks like the help and `{}` will be it stdout. The error message is in `stderr`.

Comment (samdoran) [reply lib/ansible/cli/galaxy.py:1497]: So the fact that the help is in the `stdout` could cause issues. If the help went to `stderr`, that would be better in this case, but I don't think we should change that.

Comment (shanemcd) [reply lib/ansible/cli/galaxy.py:1497]: Ok. Let's roll with it how it is now.

File: lib/ansible/cli/galaxy.py (hunk 17 above)
Comment (shanemcd) [reply lib/ansible/cli/galaxy.py:306]: Talked with Sam and he said this is not a blocker. Just going to leave in this way.

**Review by samdoran — APPROVED** [review]

**Comment by shanemcd** [comment]
Quick, someone push the button before I break the CI again!
